### PR TITLE
[Tiered Caching] Segmented cache changes 

### LINKED ISSUE #13989: [Tiered Caching] Performance improvement for TieredCaching
### Is your feature request related to a problem? Please describe

The current implementation of TieredCaching.computeIfAbsent() is performing poorly which was discovered while running some performance benchmarks. Before we delve into the root cause, lets me elaborate on the current implementation.

**Current implementation:**

As of now TieredCaching implementation has onheap and disk tier which are individually thread safe. But we need to ensure that the TieredCache as a whole is thread safe and doesn't allow any key to be cached onto both tiers.

So we implemented a single global read/write lock([link](https://github.com/opensearch-project/OpenSearch/blob/main/modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java#L80-L82)) which ensures that any write operation is performed on TieredCache by only one thread at a time and allows multiple readers to access any cache via read lock.

**Performance Issues with currrent approach:**

1. Considering we have a **single write/read lock**, it makes the TieredCache a single segmented cache and doesn't provide write throughout benefits of a segmented cache(like the underlying onheap cache which has 256 segments). Due to this only one thread can write to cache at a time and impacts write/read throughput.
2. In case of any cache miss, we are **recomputing the query under the write lock**([link](https://github.com/opensearch-project/OpenSearch/blob/main/modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java#L195)) which makes the other thread wait and makes the performance really worse. 

### Describe the solution you'd like

To address above performance issues:

#### 1. Firstly we need to move out the query recompute part(during cache miss) outside the write lock in computeIfAbsent

This can be probably achieved by below(reusing Cache.java logic).

- Create a concurrent hashmap of `Map<Key, CompletableFuture>`. This will temporally hold the queries for short duration.
- Multiple requests for same key A lands in computeIfAbsent. key A is not present in Cache yet.
- Only one thread will succeed putting a future object in the map for key A and rest will reuse the same future.
- The thread will succeeded putting a future in the map will load the query. Rest of the threads will wait via `future.get()`.
- Once query is loaded, the desired thread will put the item onto onHeap cache via `put` call. Rest of the threads will return the value.
- Remove the desired key from the map.

**2. Avoid using a single read/write lock**

Possible solutions:

- **Segmented cache**: Instead of using a single write/read local and a single segment, we can use a segmented tiered cache to improve read/write performance.
- **Lock per key**: Considering lock usage in TieredCache is to ensure that no same key lands onto both caches, we can considering creating locks per key and then delete them once the request completes. 
     -  There might be some overhead in creating locks frequently. (Will run performance tests)
     - One pros of this is that we don't really need to move the query recompute part of write lock considering other requests for different key will not be blocked.
 

### Related component

Search:Performance

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sgup432**
### Description
This is a change for making tiered cache segmented. As of today, we only had a single segment which was not as performant. This change exposes a setting to set segments for tiered cache.

### Related Issues
Resolves #13989

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: modules/cache-common/src/test/java/org/opensearch/cache/common/tier/MockDiskCache.java [hunk 1]
```diff
@@ -86,8 +87,10 @@ public V computeIfAbsent(ICacheKey<K> key, LoadAwareCacheLoader<ICacheKey<K>, V>
 
     @Override
     public void invalidate(ICacheKey<K> key) {
-        removalListener.onRemoval(new RemovalNotification<>(key, cache.get(key), RemovalReason.INVALIDATED));
-        this.cache.remove(key);
+        V value = this.cache.remove(key);
+        if (value != null) {
+            removalListener.onRemoval(new RemovalNotification<>(key, cache.get(key), RemovalReason.INVALIDATED));
+        }
```
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/test/java/org/opensearch/cache/common/tier/MockDiskCache.java:93]: Yeah a bug fix but this class is only used alongwise TieredCache UT/IT to plugin a mock disk cache implementation. This was found out during segmented cache tests, so fixed it here itself.

File: server/src/main/java/org/opensearch/common/cache/Cache.java [hunk 2]
```diff
@@ -115,8 +115,17 @@ public class Cache<K, V> {
     // the removal callback
     private RemovalListener<K, V> removalListener = notification -> {};
 
-    // use CacheBuilder to construct
-    Cache() {}
+    private int numberOfSegments = 256;
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/Cache.java:118]: Makes sense. Will change

File: server/src/main/java/org/opensearch/common/cache/Cache.java [hunk 3]
```diff
@@ -151,6 +160,13 @@ void setMaximumWeight(long maximumWeight) {
         this.maximumWeight = maximumWeight;
     }
 
+    void setNumberOfSegments(int numberOfSegments) {
+        if (numberOfSegments < 0) {
+            throw new IllegalArgumentException("numberOfSegments < 0");
+        }
+        this.numberOfSegments = numberOfSegments;
+    }
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/Cache.java:168]: Yeah this one is not needed. Will remove.

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsIT.java [hunk 4]
```diff
@@ -238,7 +237,7 @@ public void testInvalidLevelsAreIgnored() throws Exception {
         internalCluster().startNodes(
             1,
             Settings.builder()
-                .put(TieredSpilloverCacheIT.defaultSettings(HEAP_CACHE_SIZE_STRING))
+                .put(TieredSpilloverCacheIT.defaultSettings(HEAP_CACHE_SIZE_STRING, TieredSpilloverCacheIT.getNumberOfSegments()))
```
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsIT.java:240]: `getNumberOfSegments` is only used by caching related classes (IT/UT), so putting inside a common ```OpenSearchIntegTestCase``` did not make much sense to me as it will not be used elsewhere. Let me know what you think.

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java [hunk 5]
```diff
@@ -149,7 +156,7 @@ public void testSanityChecksWithIndicesRequestCache() throws InterruptedExceptio
 
     public void testWithDynamicTookTimePolicy() throws Exception {
         int onHeapCacheSizeInBytes = 2000;
-        internalCluster().startNode(Settings.builder().put(defaultSettings(onHeapCacheSizeInBytes + "b")).build());
+        internalCluster().startNode(Settings.builder().put(defaultSettings(onHeapCacheSizeInBytes + "b", 1)).build());
```
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java:159]: Doing this in a segmented way was kinda tough/impossible here. As we wanted to forcefully evict items from onheap cache to check whether disk cache policy works or not and also check validate specific stats (like number of items, misses, hits etc). Considering we don't have much visibility as to which key will land up in which segment(in IT) and related stats, I just kept the segment here as 1.

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java [hunk 6]
```diff
@@ -426,10 +435,13 @@ public void testWithExplicitCacheClear() throws Exception {
     }
 
     public void testWithDynamicDiskCacheSetting() throws Exception {
-        int onHeapCacheSizeInBytes = 10; // Keep it low so that all items are cached onto disk.
+        int numberOfSegments = getNumberOfSegments();
+        int onHeapCacheSizeInBytes = randomIntBetween(numberOfSegments + 1, numberOfSegments * 2); // Keep it low so
```
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java:439]: Yes. As per entry size on cache is around ~700 bytes(had checked it manually), so this will still work 😁

File: server/src/main/java/org/opensearch/common/cache/Cache.java [hunk 7]
```diff
@@ -115,8 +115,17 @@ public class Cache<K, V> {
     // the removal callback
     private RemovalListener<K, V> removalListener = notification -> {};
 
-    // use CacheBuilder to construct
-    Cache() {}
+    private int numberOfSegments = 256;
+
+    Cache(final int numberOfSegments) {
+        if (numberOfSegments != -1) {
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/Cache.java:122]: Done.

File: server/src/main/java/org/opensearch/common/cache/Cache.java [hunk 8]
```diff
@@ -940,6 +953,6 @@ public ToLongBiFunction<K, V> getWeigher() {
     }
 
     private CacheSegment<K, V> getCacheSegment(K key) {
-        return segments[key.hashCode() & 0xff];
+        return segments[key.hashCode() & (this.numberOfSegments - 1)];
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/Cache.java:952]: Done

File: server/src/main/java/org/opensearch/common/cache/Cache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/Cache.java]: Done

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jainankitk) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java:440]: > Plus for a segment = 256 for example, per segment onHeap cache would be 512/256 ~ 2 bytes which is still pretty less and not enough to hold a single cache item

Ah, that's a good point!

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsIT.java:240]: Actually makes sense to have a common base IT class for tiered cache. Will make change.

[the remark this answers is not in the document]
Comment (jainankitk) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsIT.java]: Maybe we can duplicate the method across classes, since it is small anyway!

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/cache/common/tier/TieredSpilloverCacheIT.java:210]: I have also added a multi segment version of this test but not doing exhaustive checks from stats perspective. But checking evictions which should be enough to validate.

File: modules/cache-common/src/test/java/org/opensearch/cache/common/tier/MockDiskCache.java [hunk 9]
```diff
@@ -62,6 +62,7 @@ public void put(ICacheKey<K> key, V value) {
         if (this.cache.size() >= maxSize) { // For simplification
             this.removalListener.onRemoval(new RemovalNotification<>(key, value, RemovalReason.EVICTED));
             this.statsHolder.decrementItems(List.of());
+            return;
```
[the remark this answers is not in the document]
Comment (jainankitk) [reply modules/cache-common/src/test/java/org/opensearch/cache/common/tier/MockDiskCache.java:65]: Although, this is okay from this test perspective, it will be better to have more complete implementation maybe using something existing

File: server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java [hunk 10]
```diff
@@ -174,18 +178,37 @@ public <K, V> ICache<K, V> create(CacheConfig<K, V> config, CacheType cacheType,
...
             ICacheBuilder<K, V> builder = new Builder<K, V>().setDimensionNames(config.getDimensionNames())
                 .setStatsTrackingEnabled(statsTrackingEnabled)
-                .setMaximumWeightInBytes(((ByteSizeValue) settingList.get(MAXIMUM_SIZE_IN_BYTES_KEY).get(settings)).getBytes())
                 .setExpireAfterAccess(((TimeValue) settingList.get(EXPIRE_AFTER_ACCESS_KEY).get(settings)))
                 .setWeigher(config.getWeigher())
                 .setRemovalListener(config.getRemovalListener());
             Setting<String> cacheSettingForCacheType = CacheSettings.CACHE_TYPE_STORE_NAME.getConcreteSettingForNamespace(
                 cacheType.getSettingPrefix()
             );
+            long maxSizeInBytes = ((ByteSizeValue) settingList.get(MAXIMUM_SIZE_IN_BYTES_KEY).get(settings)).getBytes();
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:187]: I am assuming this Cache is initialized during node startup. If not, the value of `MAXIMUM_SIZE_IN_BYTES_KEY` might change during initialization. Also, I am wondering why the `NUMBER_OF_SEGMENTS` is passed via config, and `MAXIMUM_SIZE_IN_BYTES_KEY` is directly read through setting.

Comment (jainankitk) [reply server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:187]: Just saw there is `getNumberOfSegments()` method with ICacheBuilder. Should we not be using that to add it in CacheConfig?

File: server/src/main/java/org/opensearch/common/cache/store/config/CacheConfig.java [hunk 11]
```diff
@@ -138,6 +147,18 @@ public boolean getStatsTrackingEnabled() {
         return statsTrackingEnabled;
     }
 
+    public String getStoragePath() {
+        return storagePath;
+    }
+
+    public int getNumberOfSegments() {
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/common/cache/store/config/CacheConfig.java:154]: nit: getSegmentCount might be clearer?

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheSettings.java [hunk 12]
```diff
@@ -50,6 +56,18 @@ public class TieredSpilloverCacheSettings {
...
     );
 
+    /**
+     * Setting defining the number of segments within Tiered cache
+     */
+    public static final Setting.AffixSetting<Integer> TIERED_SPILLOVER_SEGMENTS = Setting.suffixKeySetting(
+        TieredSpilloverCache.TieredSpilloverCacheFactory.TIERED_SPILLOVER_CACHE_NAME + ".segments",
+        (key) -> Setting.intSetting(key, 16, 1, k -> {
+            if (!VALID_SEGMENT_NUMBER_LIST.contains(k)) {
+                throw new IllegalArgumentException(INVALID_SEGMENT_NUMBER_EXCEPTION_MESSAGE);
```
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheSettings.java:66]: Makes sense, will change

File: server/src/main/java/org/opensearch/common/cache/settings/CacheSettings.java [hunk 13]
```diff
@@ -12,12 +12,16 @@
...
 
+import java.util.List;
+
 /**
  * Settings related to cache.
  */
 @ExperimentalApi
 public class CacheSettings {
 
+    public static final List<Integer> VALID_SEGMENT_NUMBER_LIST = List.of(1, 2, 4, 8, 16, 32, 64, 128, 256);
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/settings/CacheSettings.java:23]: Sure, will do both.

File: server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java [hunk 14]
```diff
@@ -174,18 +178,37 @@ public <K, V> ICache<K, V> create(CacheConfig<K, V> config, CacheType cacheType,
...
+                if (perSegmentSizeInBytes <= 0) {
+                    throw new IllegalArgumentException("Per segment size for opensearch onHeap cache should be greater than 0");
+                }
+                builder.setMaximumWeightInBytes(perSegmentSizeInBytes);
+                // In case this is the last segment, assign the remainder of bytes accordingly
+                if (config.getSegmentNumber() == config.getNumberOfSegments()) {
+                    if (maxSizeInBytes % config.getNumberOfSegments() != 0) {
+                        builder.setMaximumWeightInBytes(perSegmentSizeInBytes + maxSizeInBytes % config.getNumberOfSegments());
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:200]: Yeah I guess allocating 15 bytes(in case segment count is 16) to the last one does not add any real value. Will change it.

File: server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java (hunk 10 above)
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:187]: Yeah this might seem a bit confusing, let me try explaining this if I understood your question correctly. 
>Also, I am wondering why the NUMBER_OF_SEGMENTS is passed via config, and MAXIMUM_SIZE_IN_BYTES_KEY is directly read through setting.

CacheConfig is a way to pass certain parameters from core to plugin and vice versa. Each cache maintains its own max size setting(`MAXIMUM_SIZE_IN_BYTES_KEY`) which can configured by the user. So onHeap has its own, disk cache has its own, where tieredCache doesn't as it is just a composition of individual caches.

In case when a user wants to use tiered cache with X segments, tiered cache will need to create X heap/disk cache objects and need to pass this segmentCount to each respective cache implementation(so that each cache allocates each segment `maxSize/X`) which is achieved via CacheConfig. TieredCache isn't aware of each cache's `maxSize` value beforehand as it is managed by each cache's settings, so this one is fetched from setting.

Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:187]: >Just saw there is getNumberOfSegments() method with ICacheBuilder. Should we not be using that to add it in CacheConfig?

As mentioned above, CacheConfig is used here to pass numberOfSegments from tieredCache to respective cache implementation either sitting core(like this one) or plugin(like ehcache diskCache). And then cache reads it(via factory) and forms its own `ICacheBuilder` to create a respective object. So we can't use ICacheBuilder to communicate this segmentNumber/count.

File: server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java [hunk 15]
```diff
@@ -63,6 +63,10 @@ public OpenSearchOnHeapCache(Builder<K, V> builder) {
         if (builder.getExpireAfterAcess() != null) {
             cacheBuilder.setExpireAfterAccess(builder.getExpireAfterAcess());
         }
+        if (builder.getNumberOfSegments() > 0) {
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:66]: Actually even when we define segment=1 at tieredCache level, this(underlying onHeap cache) should still be set as 1.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 16]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
-        cacheListMap.put(diskCache, new TierInfo(isDiskCacheEnabled, TIER_DIMENSION_VALUE_DISK));
-        this.caches = Collections.synchronizedMap(cacheListMap);
-
         this.dimensionNames = builder.cacheConfig.getDimensionNames();
         // Pass "tier" as the innermost dimension name, in addition to whatever dimensions are specified for the cache as a whole
         this.statsHolder = new TieredSpilloverCacheStatsHolder(dimensionNames, isDiskCacheEnabled);
+        this.tieredSpilloverCacheSegments = new TieredSpilloverCacheSegment[this.numberOfSegments];
+        for (int i = 0; i < numberOfSegments; i++) {
+            tieredSpilloverCacheSegments[i] = new TieredSpilloverCacheSegment<K, V>(builder, statsHolder, i + 1, this.numberOfSegments);
+        }
```
Comment (jainankitk) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:110]: Although not part of this PR, just curious why do we need settingsUpdateConsumer for below setting. As per my understand, it is anyways not dynamic, right?
```
builder.cacheConfig.getClusterSettings()
            .addSettingsUpdateConsumer(DISK_CACHE_ENABLED_SETTING_MAP.get(builder.cacheType), this::enableDisableDiskCache);
```

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 17]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
-    private final ICache<K, V> diskCache;
-    private final ICache<K, V> onHeapCache;
-
-    // Removal listeners for the individual tiers
-    private final RemovalListener<ICacheKey<K>, V> onDiskRemovalListener;
-    private final RemovalListener<ICacheKey<K>, V> onHeapRemovalListener;
-
-    // Removal listener from the spillover cache as a whole
-    private final RemovalListener<ICacheKey<K>, V> removalListener;
+    static final String NUMBER_OF_SEGMENTS_ZERO_EXCEPTION_MESSAGE = "Number of segments cannot be less than or equal " + "to zero";
```
[the remark this answers is not in the document]
Comment (jainankitk) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:75]: Also, `Segment Count cannot be less than one` might be clearer than `less than equal to zero`

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 18]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
 
     // In future we want to just read the stats from the individual tiers' statsHolder objects, but this isn't
     // possible right now because of the way computeIfAbsent is implemented.
     private final TieredSpilloverCacheStatsHolder statsHolder;
     private ToLongBiFunction<ICacheKey<K>, V> weigher;
```
Comment (jainankitk) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:80]: nit: Do we still need the weigher here?

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 19]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
     private final List<String> dimensionNames;
-    ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
-    ReleasableLock readLock = new ReleasableLock(readWriteLock.readLock());
-    ReleasableLock writeLock = new ReleasableLock(readWriteLock.writeLock());
-    /**
-     * Maintains caching tiers in ascending order of cache latency.
-     */
-    private final Map<ICache<K, V>, TierInfo> caches;
+
     private final List<Predicate<V>> policies;
```
Comment (jainankitk) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:83]: nit: Do we still need the policies here?

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 20]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
+            } else {
+                // Put it inside desired tier.
+                try (ReleasableLock ignore = writeLock.acquire()) {
+                    for (Map.Entry<ICache<K, V>, TierInfo> entry : this.caches.entrySet()) {
+                        if (cacheValueTuple.v2().equals(entry.getValue().tierName)) {
+                            entry.getKey().put(key, value);
+                        }
+                    }
+                    updateStatsOnPut(cacheValueTuple.v2(), key, value);
+                }
```
Comment (jainankitk) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:252]: The behavior for put seems to be different than before. SegmentedCache should not be the reason for that I guess.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 21]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
-                statsHolder.incrementHits(heapDimensionValues);
-            } else if (cacheValueTuple.v2().equals(TIER_DIMENSION_VALUE_DISK)) {
-                // Miss for the heap tier, hit for the disk tier
-                statsHolder.incrementMisses(heapDimensionValues);
-                statsHolder.incrementHits(diskDimensionValues);
+        }
+
+        @Override
+        public void invalidate(ICacheKey<K> key) {
+            for (Map.Entry<ICache<K, V>, TierInfo> cacheEntry : caches.entrySet()) {
```
Comment (jainankitk) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:360]: Does the below comment no longer apply to invalidate logic?

```
// We are trying to invalidate the key from all caches though it would be present in only of them.
        // Doing this as we don't know where it is located. We could do a get from both and check that, but what will
        // also trigger a hit/miss listener event, so ignoring it for now.
        // We don't update stats here, as this is handled by the removal listeners for the tiers.
```

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 22]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
+        }
+
+        /**
+         * Get a value from the tiered cache, and the name of the tier it was found in.
+         * @param captureStats Whether to record hits/misses for this call of the function
+         * @return A tuple of the value and the name of the tier it was found in.
+         */
+        private Function<ICacheKey<K>, Tuple<V, String>> getValueFromTieredCache(boolean captureStats, boolean forceCheck) {
+            return key -> {
+                try (ReleasableLock ignore = readLock.acquire()) {
```
Comment (jainankitk) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:465]: Added this new forceCheck parameter without updating the documentation. We should avoid such changes as part of large refactoring. Can you explain the purpose of this new parameter?

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java (hunk 16 above)
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:110]: This is dynamic, we use this to enable/disable disk cache dynamically.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java (hunk 20 above)
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:252]: Yeah this is unrelated to segmented cache. Earlier` put()` logic was incorrect, as it tries to put the value without checking whether its present in the cache and replacing it accordingly. So did club this change considering it was not substantial

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java (hunk 19 above)
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:83]: Not needed, will remove.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java (hunk 18 above)
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:80]: Not needed, will remove.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java (hunk 22 above)
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:465]: My bad, seems like this change is not needed and made it here due to some sample changes I was trying. Will remove.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 23]
```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
...
+                    .setKeyType(builder.cacheConfig.getKeyType())
+                    .setValueType(builder.cacheConfig.getValueType())
+                    .setSettings(builder.cacheConfig.getSettings())
+                    .setWeigher(builder.cacheConfig.getWeigher())
+                    .setDimensionNames(builder.cacheConfig.getDimensionNames())
+                    .setMaxSizeInBytes(builder.cacheConfig.getMaxSizeInBytes())
+                    .setExpireAfterAccess(builder.cacheConfig.getExpireAfterAccess())
+                    .setClusterSettings(builder.cacheConfig.getClusterSettings())
+                    .setNumberOfSegments(numberOfSegments)
+                    .setSegmentNumber(segmentNumber)
```
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:174]: It does make sense to have a size level setting for TieredCache, also considering we are introducing segmentCount setting at TieredCache now. 
As discussed, I will also wait for @sohami review/reply on this and will make changes accordingly.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java (hunk 21 above)
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:360]: Yeah this comment was mostly stale. It was added initially when we were not maintaining cache stats explicitly within TieredCache and using listeners for that.
Though some parts still hold true ie why are we invalidating from both/all caches and the reason being for simplicity.    So removed it altogether. Will change this logic separately if needed.

File: server/src/main/java/org/opensearch/common/cache/CacheBuilder.java [hunk 24]
```diff
@@ -48,13 +50,22 @@ public class CacheBuilder<K, V> {
...
 
     public static <K, V> CacheBuilder<K, V> builder() {
         return new CacheBuilder<>();
     }
 
     private CacheBuilder() {}
 
+    public CacheBuilder<K, V> setNumberOfSegments(int numberOfSegments) {
+        if (!VALID_SEGMENT_NUMBER_LIST.contains(numberOfSegments)) {
+            throw new IllegalArgumentException("Number of segments for cache should be a power of two up-to 256");
```
[the remark this answers is not in the document]
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/CacheBuilder.java]: Using a common exception message now

File: server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java (hunk 10 above)
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:187]: This is also related to whether we should use maxSize at tieredCache level as well. Will wait on that for us to decide.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:171]: As discussed, it makes sense to have a size level setting at TieredCache level. Will make the change. And remove these confusing segmentCount/number settings.

[the remark this answers is not in the document]
Comment (sgup432) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java]: Added settings for size for respective cache tier. @jainankitk @sohami

File: server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java (hunk 10 above)
Comment (sgup432) [reply server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java:193]: Seems like this is related to https://github.com/opensearch-project/OpenSearch/pull/16047#discussion_r1785498485. Resolving this one for now, reopen if any issues.

**Review by jainankitk — APPROVED** [review]
